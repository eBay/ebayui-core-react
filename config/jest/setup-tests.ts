import '@testing-library/jest-dom'

declare var global: typeof globalThis;

if (typeof window.URL.createObjectURL === 'undefined') {
  Object.defineProperty(window.URL, 'createObjectURL', { value: () => {} });
}
