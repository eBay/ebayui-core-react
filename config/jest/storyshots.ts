import { render } from "@testing-library/react";
import initStoryshotsAddon from "@storybook/addon-storyshots";
import type { StoryshotsOptions } from "@storybook/addon-storyshots/dist/ts3.9/api/StoryshotsOptions";

const reactTestingLibrarySerializer = {
    print: (val, serialize, indent) => serialize(val.asFragment()),
    test: (val) => val && val.hasOwnProperty("container"),
};

export function initStoryshots(config: StoryshotsOptions) {
    initStoryshotsAddon({
        ...config,
        renderer: render,
        snapshotSerializers: [reactTestingLibrarySerializer],
    });
}
