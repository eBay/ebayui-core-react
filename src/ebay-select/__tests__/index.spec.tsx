import React from 'react'
import { fireEvent, render } from '@testing-library/react';
import { initStoryshots } from '../../../config/jest/storyshots';
import { EbaySelect, EbaySelectOption } from '../index'

initStoryshots({
    config: ({ configure }) =>
        configure(() => {
            require('./index.stories')
        }, module)
})

describe('<EbaySelect>', () => {
    describe('on render', () => {
        it('should pass a proper className to a wrapper element', () => {
            const wrapper = render(
                <EbaySelect name="form-select" value="1" className="blueText">
                    <EbaySelectOption value="1">Option 1</EbaySelectOption>
                    <EbaySelectOption value="2">Option 2</EbaySelectOption>
                    <EbaySelectOption value="3">Option 3</EbaySelectOption>
                </EbaySelect>)
            expect(wrapper.container.querySelectorAll('span')[0]).toHaveClass('blueText')
        })

        it('should pass a `disabled` prop to a `select` element', () => {
            const wrapper = render(
                <EbaySelect name="form-select" value="1" disabled>
                    <EbaySelectOption value="1">Option 1</EbaySelectOption>
                    <EbaySelectOption value="2">Option 2</EbaySelectOption>
                    <EbaySelectOption value="3">Option 3</EbaySelectOption>
                </EbaySelect>
            )
            expect(wrapper.getByRole('combobox')).toHaveAttribute('disabled', '')
        })
    })

    describe('on select-dropdown change', () => {
        it('should fire onChange event', () => {
            const spy = jest.fn()
            const wrapper = render(
                <EbaySelect name="foo" value="1" onChange={spy} >
                    <EbaySelectOption value="1">Option 1</EbaySelectOption>
                    <EbaySelectOption value="2">Option 2</EbaySelectOption>
                    <EbaySelectOption value="3">Option 3</EbaySelectOption>
                </EbaySelect>
            )
            simulateSelectChange(wrapper, '3', 2)

            expect(spy).toBeCalledTimes(1)
            expect(spy).toBeCalledWith(expect.anything(), 2, '3')
        })
    })

    describe('on uncontrolled component', () => {
        let wrapper
        let spy

        beforeEach(() => {
            spy = jest.fn()
            wrapper = render(
                <EbaySelect name="foo" defaultValue="2" onChange={spy} >
                    <EbaySelectOption value="1">Option 1</EbaySelectOption>
                    <EbaySelectOption value="2">Option 2</EbaySelectOption>
                    <EbaySelectOption value="3">Option 3</EbaySelectOption>
                </EbaySelect>
            )
        })

        it('should set initial selected value', () => {
            expect(wrapper.getByRole('combobox')).toHaveValue('2')
        })

        it('should automatically render the newly selected value when select changes', () => {
            simulateSelectChange(wrapper, '3', 2)

            expect(wrapper.getByRole('combobox')).toHaveValue('3')
        })
    })

    describe('in case for controlled component', () => {
        let wrapper
        let spy

        beforeEach(() => {
            spy = jest.fn()
            wrapper = render(
                <EbaySelect name="foo" value="2" onChange={spy} >
                    <EbaySelectOption value="1">Option 1</EbaySelectOption>
                    <EbaySelectOption value="2">Option 2</EbaySelectOption>
                    <EbaySelectOption value="3">Option 3</EbaySelectOption>
                </EbaySelect>
            )
        })

        it('should set initial selected value', () => {
            expect(wrapper.getByRole('combobox')).toHaveValue('2')
        })

        it('should not automatically update the newly selected value when select changes', () => {
            simulateSelectChange(wrapper, '3', 2)

            expect(wrapper.getByRole('combobox')).toHaveValue('2')
        })
    })
})

function simulateSelectChange(wrapper, selectedValue, selectedIndex) {
    fireEvent.change(wrapper.getByRole('combobox'), {
        target: {
            value: selectedValue,
            selectedIndex: selectedIndex
        }
    })
}
