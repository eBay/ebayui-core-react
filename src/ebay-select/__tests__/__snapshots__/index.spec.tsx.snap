// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots ebay-select Basic 1`] = `
<DocumentFragment>
  <span
    class="select"
  >
    <select
      name="formSelect"
    >
      <option
        value="1"
      >
        Option 1
      </option>
      <option
        value="2"
      >
        Option 2
      </option>
      <option
        value="3"
      >
        Option 3
      </option>
    </select>
    <svg
      aria-hidden="true"
      class="icon icon--dropdown"
      focusable="false"
      height="8"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        xlink:href="#icon-dropdown"
      />
    </svg>
  </span>
</DocumentFragment>
`;

exports[`Storyshots ebay-select Borderless select 1`] = `
<DocumentFragment>
  <span
    class="select select--borderless"
  >
    <select
      name="formSelect"
    >
      <option
        value="1"
      >
        Option 1
      </option>
      <option
        value="2"
      >
        Option 2
      </option>
      <option
        value="3"
      >
        Option 3
      </option>
    </select>
    <svg
      aria-hidden="true"
      class="icon icon--dropdown"
      focusable="false"
      height="8"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        xlink:href="#icon-dropdown"
      />
    </svg>
  </span>
</DocumentFragment>
`;

exports[`Storyshots ebay-select Controlled component 1`] = `
<DocumentFragment>
  <div
    style="width: 500px;"
  >
    <div
      style="display: flex; align-items: center;"
    >
      <span
        class="select"
      >
        <select
          name="selectCountry"
        >
          <option
            value="DE"
          >
            DE
          </option>
          <option
            value="US"
          >
            US
          </option>
          <option
            value="UK"
          >
            UK
          </option>
          <option
            value="AU"
          >
            AU
          </option>
        </select>
        <svg
          aria-hidden="true"
          class="icon icon--dropdown"
          focusable="false"
          height="8"
          width="8"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-dropdown"
          />
        </svg>
      </span>
      <div
        style="margin: 0px 2rem;"
      >
        Current selected: 
        <strong>
          UK
        </strong>
      </div>
      <button
        class="btn btn--secondary"
      >
        Set selected to US
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Storyshots ebay-select Disabled select 1`] = `
<DocumentFragment>
  <span
    class="select customclass"
  >
    <select
      disabled=""
      name="formSelect"
    >
      <option
        value="1"
      >
        Option 1
      </option>
      <option
        value="2"
      >
        Option 2
      </option>
    </select>
    <svg
      aria-hidden="true"
      class="icon icon--dropdown"
      focusable="false"
      height="8"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        xlink:href="#icon-dropdown"
      />
    </svg>
  </span>
</DocumentFragment>
`;

exports[`Storyshots ebay-select Floating label 1`] = `
<DocumentFragment>
  <span
    class="floating-label"
  >
    <label
      class="floating-label__label"
    >
      Label
    </label>
    <span
      class="select"
    >
      <select
        name="formSelect"
      >
        <option
          value=""
        />
        <option
          value="1"
        >
          Option 1
        </option>
        <option
          value="2"
        >
          Option 2
        </option>
        <option
          value="3"
        >
          Option 3
        </option>
      </select>
      <svg
        aria-hidden="true"
        class="icon icon--dropdown"
        focusable="false"
        height="8"
        width="8"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlink:href="#icon-dropdown"
        />
      </svg>
    </span>
  </span>
</DocumentFragment>
`;

exports[`Storyshots ebay-select Grouped options 1`] = `
<DocumentFragment>
  <span
    class="select"
  >
    <select
      name="formSelect"
    >
      <option
        value="pre"
      >
        Option PRE
      </option>
      <optgroup
        label="Group 1"
      >
        <option
          value="1"
        >
          Option 1[Group 1]
        </option>
        <option
          value="2"
        >
          Option 2[Group 1]
        </option>
        <option
          value="3"
        >
          Option 3[Group 1]
        </option>
      </optgroup>
      <option
        value="mid"
      >
        Option MID
      </option>
      <optgroup
        label="Group 2"
      >
        <option
          value="4"
        >
          Option 4[Group 2]
        </option>
        <option
          value="5"
        >
          Option 5[Group 2]
        </option>
        <option
          value="6"
        >
          Option 6[Group 2]
        </option>
      </optgroup>
      <optgroup
        label="Group 3"
      >
        <option
          value="7"
        >
          Option 7[Group 3]
        </option>
        <option
          value="8"
        >
          Option 8[Group 3]
        </option>
        <option
          value="9"
        >
          Option 9[Group 3]
        </option>
      </optgroup>
      <option
        value="post"
      >
        Option POST
      </option>
    </select>
    <svg
      aria-hidden="true"
      class="icon icon--dropdown"
      focusable="false"
      height="8"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        xlink:href="#icon-dropdown"
      />
    </svg>
  </span>
</DocumentFragment>
`;

exports[`Storyshots ebay-select Invalid floating label select 1`] = `
<DocumentFragment>
  <span
    class="floating-label"
  >
    <label
      class="floating-label__label floating-label__label--invalid"
    >
      Invalid label
    </label>
    <span
      class="select"
    >
      <select
        aria-invalid="true"
        aria-label="Please select a option"
        name="formSelect"
      >
        <option
          value="1"
        >
          Option 1
        </option>
        <option
          value="2"
        >
          Option 2
        </option>
        <option
          value="3"
        >
          Option 3
        </option>
      </select>
      <svg
        aria-hidden="true"
        class="icon icon--dropdown"
        focusable="false"
        height="8"
        width="8"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlink:href="#icon-dropdown"
        />
      </svg>
    </span>
  </span>
</DocumentFragment>
`;

exports[`Storyshots ebay-select Invalid select 1`] = `
<DocumentFragment>
  <span
    class="select"
  >
    <select
      aria-invalid="true"
      aria-label="Please select a option"
      name="formSelect"
    >
      <option
        value="1"
      >
        Option 1
      </option>
      <option
        value="2"
      >
        Option 2
      </option>
      <option
        value="3"
      >
        Option 3
      </option>
    </select>
    <svg
      aria-hidden="true"
      class="icon icon--dropdown"
      focusable="false"
      height="8"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        xlink:href="#icon-dropdown"
      />
    </svg>
  </span>
</DocumentFragment>
`;

exports[`Storyshots ebay-select Large Select 1`] = `
<DocumentFragment>
  <span
    class="floating-label floating-label--large"
  >
    <label
      class="floating-label__label"
    >
      Label
    </label>
    <span
      class="select select--large"
    >
      <select
        name="formSelect"
      >
        <option
          value=""
        />
        <option
          value="1"
        >
          Option 1
        </option>
        <option
          value="2"
        >
          Option 2
        </option>
        <option
          value="3"
        >
          Option 3
        </option>
      </select>
      <svg
        aria-hidden="true"
        class="icon icon--dropdown"
        focusable="false"
        height="8"
        width="8"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlink:href="#icon-dropdown"
        />
      </svg>
    </span>
  </span>
</DocumentFragment>
`;

exports[`Storyshots ebay-select Options class 1`] = `
<DocumentFragment>
  <span
    class="select"
  >
    <select
      name="formSelect"
    >
      <option
        value="1"
      >
        Option 1
      </option>
      <option
        value="2"
      >
        Option 2
      </option>
      <option
        value="3"
      >
        Option 3
      </option>
    </select>
    <svg
      aria-hidden="true"
      class="icon icon--dropdown"
      focusable="false"
      height="8"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        xlink:href="#icon-dropdown"
      />
    </svg>
  </span>
</DocumentFragment>
`;

exports[`Storyshots ebay-select Uncontrolled component with defaultValue 1`] = `
<DocumentFragment>
  <span
    class="select"
  >
    <select
      name="formSelect"
    >
      <option
        value="1"
      >
        Option 1
      </option>
      <option
        value="2"
      >
        Option 2
      </option>
      <option
        value="3"
      >
        Option 3
      </option>
    </select>
    <svg
      aria-hidden="true"
      class="icon icon--dropdown"
      focusable="false"
      height="8"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        xlink:href="#icon-dropdown"
      />
    </svg>
  </span>
</DocumentFragment>
`;
